concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -116,30 +125,122 @@ sectionHeaderLength :: Word64
+
+-- |Data type used to represent a line with genesis block hash in the block index file.
+type BlockIndexGenesisHash = BlockHash
+
+-- |Data type used to represent a line with chunk information in the block index file.
+data BlockIndexChunkInfo = BlockIndexChunkInfo
+    { filename :: T.Text,
+      genesisIndex :: GenesisIndex,
+      blockHeightFirst :: BlockHeight,
+      blockHeightLast :: BlockHeight
+    }
+    deriving (Show)
+
+-- |Data type used to represent the block index file.
+type BlockIndex = Seq (BlockIndexGenesisHash, Seq BlockIndexChunkInfo)
+
+-- |Parse a line of the block index file containing block genesis data.
+parseGenesisDataLine :: AP.Parser BlockIndexGenesisHash
+parseGenesisDataLine = do
+    _ <- AP.string (T.pack "# genesis hash ")
+    hash <- AP.count 64 $ AP.satisfy isHexDigit
+    AP.skip AP.isEndOfLine
+    return $ read hash
+
+-- |Parse a line of the block index file containing chunk information.
+parseChunkLine :: AP.Parser BlockIndexChunkInfo
+parseChunkLine = do
+    filename <- AP.takeWhile1 (AP.inClass ".a-zA-Z0-9-")
+    _ <- AP.char ','
+    genesisIndex <- AP.decimal
+    _ <- AP.char ','
+    blockHeightStart <- AP.decimal
+    _ <- AP.char ','
+    blockHeightEnd <- AP.decimal
+    AP.skip AP.isEndOfLine
+    return $ BlockIndexChunkInfo filename genesisIndex blockHeightStart blockHeightEnd
+
+-- | Parse a block index file.
+parseBlockIndexFile :: AP.Parser BlockIndex
+parseBlockIndexFile = do
+    atEnd <- AP.atEnd
+    if not atEnd
+        then do
+            list <-
+                AP.manyTill
+                    ( do
+                        genesisData <- parseGenesisDataLine
+                        chunks <- AP.many' parseChunkLine
+                        return (genesisData, fromList chunks)
+                    )
+                    AP.endOfInput
+            return $ fromList list
+        else return Empty
+
+-- |Show the contents of a block index file. The string can be parsed
+-- into a corresponding @BlockIndex@ using `parseBlockIndexFile`.
+showBlockIndexFile :: BlockIndex -> String
+showBlockIndexFile = concatMap showSection
+  where
+    showSection :: (BlockIndexGenesisHash, Seq BlockIndexChunkInfo) -> String
+    showSection (gd, chunks) = showGenesisDataLine gd <> concatMap showChunkLine chunks
+    showGenesisDataLine :: BlockIndexGenesisHash -> String
+    showGenesisDataLine gd = "# genesis hash " <> show gd <> "\n"
+    showChunkLine :: BlockIndexChunkInfo -> String
+    showChunkLine (BlockIndexChunkInfo{..}) =
+        T.unpack filename
+            <> ","
+            <> show genesisIndex
+            <> ","
+            <> show blockHeightFirst
+            <> ","
+            <> show blockHeightLast
+            <> "\n"
+
+-- |Normalize the block index by eliminating redundant genesis hash lines.
+-- This is necessary since older versions of the database exporter may
+-- insert superfluous (i.e. repeated) lines containing the same genesis
+-- hashes.
```

Comment:
I think the comment here does not really reflect how this function is used. It is used upon output.

If only the input is the problem then the parser could already handle this to always return a normalized block index. Why is that not enough?

Reply:
Specifically this is used in output here https://github.com/Concordium/concordium-node/pull/635/files#diff-64912044dc5a6fbbce28889ab7e8ab320e77a1dabaa2246ca15cc18b853abf81R322